powersimdata/input/changes/electrification.py

```diff
@@ -0,0 +1,121 @@
+import copy
+from dataclasses import dataclass
+from typing import Dict
+
+
+def _check_scale_factors(scale_factors):
+    """Validate schema of scale factors dict
+
+    :param dict scale_factors: see :func:`add_electrification`
+    """
+    if not all(isinstance(k, str) for k in scale_factors):
+        raise ValueError("profile name must be str")
+    if not all(isinstance(d, (int, float)) for d in scale_factors.values()):
+        raise ValueError("scale factors must be numeric")
+    vals = list(scale_factors.values())
+    if any(v < 0 for v in vals):
+        raise ValueError("scaling factor must be non negative")
+    if sum(vals) > 1:
+        raise ValueError("scaling factors must sum to between 0 and 1")
+
+
+@dataclass
+class ScaleFactors:
+    """Map technology to adoption rate
+
+    :param dict sf: a dictionary mapping tech to adoption rate
+    """
+
+    sf: Dict[str, float]
+
+    def __init__(self, sf):
+        _check_scale_factors(sf)
+        self.sf = sf
+
+    def value(self):
+        return self.sf
+
+
+@dataclass
+class AreaScaling:
+    """Map end uses to adoption rates of each technology
+
+    :param dict info: a mapping from end use (*str*) to scale factors (*dict*)
+    :raises ValueError: if info is not a dict, or any keys are not strings
+    """
+
+    end_uses: Dict[str, ScaleFactors]
+
+    def __init__(self, info):
+        if not isinstance(info, dict):
+            raise ValueError("scale factors must be a dict")
```

Comment:
What do you think of `zone/grid scaling must be a dict` since at this point `info` is a `dict` with keys being zone/grid?

Reply:
Done